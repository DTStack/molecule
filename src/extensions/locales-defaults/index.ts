import { IExtension, IContributeType } from 'mo/model/extension';

const koKR = require('./locales/ko-KR.json');
const zhCN = require('./locales/zh-CN.json');
const en = require('./locales/en.json');
const locales = [zhCN, en, koKR];

export const ExtendsLocales: IExtension = {
    id: 'ExtendsLocales',
    name: 'Extends locales',
    contributes: {
        [IContributeType.Languages]: locales,
    },
    activate() {},
    dispose() {},
};
export const BuiltInLocales = locales;
export const BuiltInId = en.id;
export const BuiltInDefault = locales.find((item) => item.id === BuiltInId);
