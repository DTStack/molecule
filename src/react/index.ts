export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
