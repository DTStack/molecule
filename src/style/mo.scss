@import 'mo/style/common';
@import 'mo/style/animation';

// ============= The Workbench parts =============
@import 'mo/components/actionBar/style';
@import 'mo/components/breadcrumb/style';
@import 'mo/components/button/style';
@import 'mo/components/checkbox/style';
@import 'mo/components/collapse/style';
@import 'mo/components/contextView/style';
@import 'mo/components/dialog/style';
@import 'mo/components/dropdown/style';
@import 'mo/components/input/style';
@import 'mo/components/list/style';
@import 'mo/components/menu/style';
@import 'mo/components/scrollable/style';
@import 'mo/components/select/style';
@import 'mo/components/tabs/style';
@import 'mo/components/toolbar/style';
@import 'mo/components/tree/style';
@import 'mo/components/search/style';
@import 'mo/components/tooltip/style';
@import 'mo/components/split/style';

// ============= The Workbench core component =============
@import 'mo/workbench/style';
@import 'mo/workbench/activityBar/style';
@import 'mo/workbench/editor/style';
@import 'mo/workbench/editor/welcome/style';
@import 'mo/workbench/menuBar/style';
@import 'mo/workbench/panel/style';
@import 'mo/workbench/settings/style';
@import 'mo/workbench/sidebar/style';
@import 'mo/workbench/statusBar/style';
@import 'mo/workbench/notification/style';
@import 'mo/workbench/sidebar/explore/style';
@import 'mo/workbench/problems/style';
@import 'mo/workbench/notification/statusBarView/style';

.#{$prefix} {
    bottom: 0;
    box-sizing: border-box;
    font-size: 13px;
    height: 100%;
    left: 0;
    margin: 0;
    outline: 0;
    overflow: hidden;
    padding: 0;
    position: absolute;
    right: 0;
    top: 0;
    user-select: none;
    width: 100%;
}

.mac {
    font-family: -apple-system, BlinkMacSystemFont, system-ui, sans-serif;
}

.center {
    text-align: center;
}

.hidden {
    display: none;
}
