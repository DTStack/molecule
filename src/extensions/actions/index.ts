import { IBottomActivityBarItem, IContributeType, type IExtension, IMenuItemProps, UniqueId } from 'mo/types';
import { concatMenu } from 'mo/utils';

import AddCursorAboveAction from './addCursorAbove';
import AddCursorBelowAction from './addCursorBelow';
import AddSelectionToNextFindMatchAction from './addSelectionToNextFindMatchAction';
import CopyAction from './copy';
import CopyLineDownAction from './copyLineDown';
import CopyLineUpAction from './copyLineUp';
import CutAction from './cut';
import ExpandSelectionAction from './expandSelection';
import FindAction from './find';
import MoveLineDownAction from './moveLineDown';
import MoveLineUpAction from './moveLineUp';
import PasteAction from './paste';
import { QuickAccessCommandAction } from './quickAccessCommandAction';
import { QuickAccessSettingsAction } from './quickAccessSettingsAction';
import QuickJumpToLineAction from './quickJumpToLineAction';
import QuickSelectLocaleAction from './quickSelectLocaleAction';
import QuickSelectThemeAction from './quickSelectThemeAction';
import QuickTogglePanelAction from './quickTogglePanelAction';
import QuickToggleSidebarAction from './quickToggleSideBarAction';
import RedoAction from './redo';
import ReplaceAction from './replace';
import SelectAllAction from './selectAll';
import SelectHighlightsAction from './selectHighlightsAction';
import ShrinkSelectionAction from './shrinkSelection';
import ToggleBlockCommentAction from './toggleBlockComment';
import ToggleLineCommentAction from './toggleLineComment';
import UndoAction from './undo';

export const ExtendsActions: IExtension = {
    id: 'ExtendsActions',
    name: 'Extend The Default Actions',
    contributes: {
        [IContributeType.Commands]: [
            UndoAction,
            RedoAction,
            SelectAllAction,
            CopyLineUpAction,
            CopyLineDownAction,
            MoveLineUpAction,
            MoveLineDownAction,
            CutAction,
            CopyAction,
            PasteAction,
            FindAction,
            ReplaceAction,
            AddCursorAboveAction,
            AddCursorBelowAction,
            ExpandSelectionAction,
            ShrinkSelectionAction,
            ToggleLineCommentAction,
            ToggleBlockCommentAction,
            AddSelectionToNextFindMatchAction,
            SelectHighlightsAction,
            QuickSelectThemeAction,
            QuickTogglePanelAction,
            QuickToggleSidebarAction,
            QuickJumpToLineAction,
            QuickAccessSettingsAction,
            QuickAccessCommandAction,
            QuickSelectLocaleAction,
        ],
    },
    activate: function (molecule): void {
        // append actions into settings' menu
        appendActionToSettingMenu(QuickAccessCommandAction);
        appendActionToSettingMenu(QuickAccessSettingsAction);
        appendActionToSettingMenu(QuickSelectThemeAction);

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_EDIT)
            .with(CutAction)
            .with(CopyAction)
            .with(PasteAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_EDIT)
            .with(FindAction)
            .with(ReplaceAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_EDIT)
            .with(ToggleLineCommentAction)
            .with(ToggleBlockCommentAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_SELECTION)
            .with(SelectAllAction)
            .with(ExpandSelectionAction)
            .with(ShrinkSelectionAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_SELECTION)
            .with(CopyLineUpAction)
            .with(CopyLineDownAction)
            .with(MoveLineUpAction)
            .with(MoveLineDownAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_SELECTION)
            .with(AddCursorAboveAction)
            .with(AddCursorBelowAction)
            .with(AddSelectionToNextFindMatchAction)
            .with(SelectHighlightsAction)
            .exhaust();

        // update menu's keybinding
        updateMenuKeybinding(QuickAccessCommandAction.ID);
        updateMenuKeybinding(QuickTogglePanelAction.ID);
        updateMenuKeybinding(QuickToggleSidebarAction.ID);
        updateMenuKeybinding(UndoAction.ID);
        updateMenuKeybinding(RedoAction.ID);
        updateMenuKeybinding(CutAction.ID);

        function appendActionToSettingMenu(ctor: { ID: string }) {
            const keybinding = molecule.action.queryGlobalKeybinding(ctor.ID);
            // Add Settings into setting's menus
            molecule.activityBar.update<IBottomActivityBarItem>(
                molecule.builtin.getState().constants.ACTIVITYBAR_ITEM_SETTING,
                (prev) => ({
                    id: prev.id,
                    contextMenu: [
                        ...(prev.contextMenu || []),
                        {
                            id: ctor.ID,
                            name: molecule.locale.localize(ctor.ID, ctor.ID),
                            keybinding: keybinding
                                ? molecule.action.convertSimpleKeybindingToString(keybinding)
                                : undefined,
                        },
                    ],
                })
            );

            molecule.activityBar.onContextMenuClick((item) => {
                if (item.id === ctor.ID) {
                    molecule.action.execute(ctor.ID);
                }
            });
        }

        function updateMenuKeybinding(id: UniqueId) {
            const keybinding = molecule.action.queryGlobalKeybinding(id);
            if (keybinding) {
                molecule.menuBar.update({
                    id,
                    keybinding: molecule.action.convertSimpleKeybindingToString(keybinding),
                });
            }
        }

        function appendActionGroupBy(parentId: UniqueId) {
            const items: IMenuItemProps[] = [];
            return new (class {
                with = (ctor: { ID: string }) => {
                    const keybinding = molecule.action.queryGlobalKeybinding(ctor.ID);

                    items.push({
                        id: ctor.ID,
                        name: molecule.locale.localize(ctor.ID, ctor.ID),
                        keybinding: keybinding
                            ? molecule.action.convertSimpleKeybindingToString(keybinding)
                            : undefined,
                    });
                    return this;
                };
                exhaust = () => {
                    const menuItems = concatMenu(molecule.menuBar.get(parentId)?.children || [], items);
                    molecule.menuBar.update(parentId, () => ({
                        children: menuItems,
                    }));
                };
            })();
        }
    },
};
