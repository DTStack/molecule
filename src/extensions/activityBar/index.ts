import { Alignment, IExtension } from 'mo/types';
import { classify, concatMenu, sortByIndex } from 'mo/utils';

import { createContextMenu, createMenuDuplicate } from '../utils';

export const ExtendsActivityBar: IExtension = {
    id: 'ExtendsActivityBar',
    name: 'Extend The Default ActivityBar',
    activate: function (molecule): void {
        molecule.activityBar.onClick((item) => {
            if (item.id === molecule.activityBar.getCurrent()) {
                molecule.layout.setSidebar((prev) => !prev);
            } else {
                molecule.layout.setSidebar(true);
                molecule.activityBar.setCurrent(item.id);
                molecule.sidebar.setCurrent(item.id);
            }
        });

        molecule.activityBar.onContextMenu((pos, activityItem) => {
            const { ACTIVITYBAR_CONTEXTMENU = [] } = molecule.builtin.getModules();
            const activityBar = molecule.activityBar.getState().data;
            const [top = [], bottom = []] = classify(
                activityBar,
                (item) => item.alignment === Alignment.top
            );
            let contextMenu = concatMenu(
                createContextMenu(top.sort(sortByIndex)),
                createContextMenu(bottom.sort(sortByIndex)),
                ACTIVITYBAR_CONTEXTMENU
            );
            const target = contextMenu.find((item) => item.id === activityItem?.id);
            if (target) {
                contextMenu = concatMenu([createMenuDuplicate(target)], contextMenu);
            }
            molecule.contextMenu.open(
                contextMenu,
                pos,
                // remark current contextMenu
                molecule.builtin.getConstants().CONTEXTMENU_ITEM_ACTIVITYBAR
            );
        });
    },
};
