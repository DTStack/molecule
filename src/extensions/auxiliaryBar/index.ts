import { IExtension } from 'mo/types';

export const ExtendsAuxiliaryBar: IExtension = {
    id: 'ExtendsAuxiliaryBar',
    name: 'Extend The Default AuxiliaryBar',
    activate: function (molecule): void {
        molecule.auxiliaryBar.onTabClick((id) => {
            molecule.auxiliaryBar.toggle(id);
        });
    },
};
