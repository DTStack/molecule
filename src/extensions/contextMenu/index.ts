import { EditorEvent, EditorGroupModel } from 'mo/models/editor';
import { EditorTreeEvent } from 'mo/models/editorTree';
import { FolderTreeEvent } from 'mo/models/folderTree';
import { ISidebarPane, SidebarEvent } from 'mo/models/sideBar';
import {
    Alignment,
    type IEditorTab,
    type IExtension,
    type IMenuItemProps,
    type TabGroup,
    type UniqueId,
} from 'mo/types';
import { getPrevOrNext, searchById, sortByIndex } from 'mo/utils';
import { TreeNodeModel } from 'mo/utils/tree';

export const ExtendsContextMenu: IExtension = {
    id: 'ExtendsContextMenu',
    name: 'Extend The Default ContextMenu',
    activate: function (molecule): void {
        molecule.contextMenu.onHide(() => {
            molecule.contextMenu.close();
        });

        molecule.contextMenu.onClick((item) => {
            const {
                MENUBAR_CONTEXTMENU_HIDE,
                ACTIVITYBAR_CONTEXTMENU_HIDE,
                STATUSBAR_CONTEXTMENU_HIDE,
                PANEL_CONTEXTMENU_HIDE,
            } = molecule.builtin.getConstants();
            const scope = molecule.contextMenu.getScope();
            switch (item.id) {
                case MENUBAR_CONTEXTMENU_HIDE: {
                    molecule.layout.setMenuBar(false);
                    break;
                }
                case ACTIVITYBAR_CONTEXTMENU_HIDE: {
                    molecule.layout.setActivityBar(false);
                    break;
                }
                case STATUSBAR_CONTEXTMENU_HIDE: {
                    molecule.layout.setStatusBar(false);
                    break;
                }
                case PANEL_CONTEXTMENU_HIDE: {
                    molecule.layout.setPanel(false);
                    break;
                }
                default: {
                    if (!scope) break;
                    // dispatch to different scope
                    if (isActivityBar(scope)) {
                        activityBarContextMenuClick(item);
                    } else if (isStatusBar(scope)) {
                        statusBarContextMenuClick(item);
                    } else if (isPanel(scope)) {
                        panelContextMenuClick(item);
                    } else if (isEditorTree(scope)) {
                        editorTreeContextMenuClick(item, scope.group, scope.tab);
                    } else if (isSidebar(scope)) {
                        sidebarContextMenuClick(item, scope.item);
                    } else if (isExplorer(scope)) {
                        explorerContextMenuClick(item);
                    } else if (isFolderTree(scope)) {
                        folderTreeContextMenuClick(item, scope.item);
                    } else if (isEditor(scope)) {
                        editorContextMenuClick(item, scope.item.tabId, scope.item.groupId);
                    }
                }
            }
        });

        molecule.contextMenu.onClick(() => {
            molecule.contextMenu.close();
        });

        function isActivityBar(scope: any): scope is string {
            return (
                typeof scope === 'string' &&
                scope === molecule.builtin.getConstants().CONTEXTMENU_ITEM_ACTIVITYBAR
            );
        }

        function activityBarContextMenuClick(item: IMenuItemProps) {
            const target = molecule.activityBar.get(item.id);
            if (!target) return;
            molecule.activityBar.toggleBar(item.id);
            if (target.alignment === Alignment.bottom) return;
            if (!target.hidden) {
                molecule.sidebar.setCurrent(item.id);
                molecule.activityBar.setCurrent(item.id);
            } else if (molecule.activityBar.getState().current === item.id) {
                const list = molecule.activityBar
                    .getState()
                    .data.filter((i) => !i.hidden && i.alignment === Alignment.top)
                    .sort(sortByIndex);
                const idx = list.findIndex(searchById(item.id));
                const next = getPrevOrNext(list, idx)?.id;
                molecule.sidebar.setCurrent(next);
                molecule.activityBar.setCurrent(next);
            }
        }

        function isStatusBar(scope: any): scope is string {
            return (
                typeof scope === 'string' &&
                scope === molecule.builtin.getConstants().CONTEXTMENU_ITEM_STATUS_BAR
            );
        }

        function statusBarContextMenuClick(item: IMenuItemProps) {
            const target = molecule.statusBar.get(item.id);
            if (target) {
                molecule.statusBar.toggle(item.id);
            }
        }

        function isPanel(scope: any): scope is string {
            return (
                typeof scope === 'string' &&
                scope === molecule.builtin.getConstants().CONTEXTMENU_ITEM_PANEL
            );
        }

        function panelContextMenuClick(item: IMenuItemProps) {
            const target = molecule.panel.get(item.id);
            if (!target) return;
            molecule.panel.toggle(item.id);
            if (!target.hidden) {
                molecule.panel.setCurrent(item.id);
            } else if (molecule.panel.getState().current === item.id) {
                const list = molecule.panel
                    .getState()
                    .data.filter((i) => !i.hidden)
                    .sort(sortByIndex);
                const idx = list.findIndex(searchById(item.id));
                const next = getPrevOrNext(list, idx)?.id;
                molecule.panel.setCurrent(next);
            }
        }

        function isEditorTree(
            scope: any
        ): scope is { name: string; group: EditorGroupModel; tab?: IEditorTab<any> } {
            return (
                typeof scope === 'object' &&
                scope.name === molecule.builtin.getConstants().CONTEXTMENU_ITEM_EDITOR_TREE
            );
        }

        function editorTreeContextMenuClick(
            item: IMenuItemProps,
            group: EditorGroupModel,
            tab?: IEditorTab<any>
        ) {
            const {
                EDITORTREE_TOOLBAR_SAVE_GROUP,
                EDITORTREE_TOOLBAR_CLOSE_GROUP,
                EDITOR_CONTEXTMENU_CLOSE,
                EDITOR_CONTEXTMENU_CLOSE_OTHERS,
                EDITOR_CONTEXTMENU_CLOSE_TO_RIGHT,
                EDITOR_CONTEXTMENU_CLOSE_TO_LEFT,
                EDITOR_CONTEXTMENU_CLOSE_ALL,
            } = molecule.builtin.getConstants();
            switch (item.id) {
                case EDITORTREE_TOOLBAR_CLOSE_GROUP:
                case EDITORTREE_TOOLBAR_SAVE_GROUP: {
                    // Redirect to editorTree's onToolbarClick event
                    molecule.editorTree.emit(EditorTreeEvent.onToolbarClick, item, group.id);
                    break;
                }
                case EDITOR_CONTEXTMENU_CLOSE: {
                    // Redirect to editor's onClose event
                    if (tab) {
                        molecule.editor.emit(EditorEvent.onCloseTab, tab.id, group.id);
                    }
                    break;
                }
                case EDITOR_CONTEXTMENU_CLOSE_OTHERS:
                    if (tab) {
                        molecule.editor.emit(EditorEvent.onCloseOther, tab.id, group.id);
                    }
                    break;
                case EDITOR_CONTEXTMENU_CLOSE_TO_RIGHT:
                    if (tab) {
                        molecule.editor.emit(EditorEvent.onCloseToRight, tab.id, group.id);
                    }
                    break;
                case EDITOR_CONTEXTMENU_CLOSE_TO_LEFT:
                    if (tab) {
                        molecule.editor.emit(EditorEvent.onCloseToLeft, tab.id, group.id);
                    }
                    break;
                case EDITOR_CONTEXTMENU_CLOSE_ALL:
                    if (tab) {
                        molecule.editor.emit(EditorEvent.onCloseAll, group.id);
                    }
                    break;
                default:
                    break;
            }
        }

        function isSidebar(scope: any): scope is { name: string; item: ISidebarPane } {
            return (
                typeof scope === 'object' &&
                scope.name === molecule.builtin.getConstants().CONTEXTMENU_ITEM_SIDEBAR
            );
        }

        function sidebarContextMenuClick(item: IMenuItemProps, pane: ISidebarPane) {
            if (pane.toolbar?.find(searchById(item.id))) {
                molecule.sidebar.emit(SidebarEvent.onToolbarClick, item, pane.id);
            }
        }

        function isExplorer(scope: any): scope is string {
            return (
                typeof scope === 'string' &&
                scope === molecule.builtin.getConstants().CONTEXTMENU_ITEM_EXPLORER
            );
        }

        function explorerContextMenuClick(item: IMenuItemProps) {
            const data = molecule.explorer.getState().data;
            if (data?.find(searchById(item.id))) {
                molecule.sidebar.emit(
                    SidebarEvent.onToolbarClick,
                    item,
                    molecule.builtin.getConstants().SIDEBAR_ITEM_EXPLORER
                );
            }
        }

        function isFolderTree(scope: any): scope is { name: string; item: TreeNodeModel<any> } {
            return (
                typeof scope === 'object' &&
                scope.name === molecule.builtin.getConstants().CONTEXTMENU_ITEM_FOLDERTREE
            );
        }

        function folderTreeContextMenuClick(item: IMenuItemProps, treeNode: TreeNodeModel<any>) {
            molecule.folderTree.emit(FolderTreeEvent.onContextMenuClick, item, treeNode);
        }

        function isEditor(scope: any): scope is { name: string; item: TabGroup } {
            return (
                typeof scope === 'object' &&
                scope.name === molecule.builtin.getConstants().CONTEXTMENU_ITEM_EDITOR
            );
        }

        function editorContextMenuClick(item: IMenuItemProps, tabId: UniqueId, groupId: UniqueId) {
            molecule.editor.emit(EditorEvent.onContextMenuClick, item, tabId, groupId);
        }
    },
};
