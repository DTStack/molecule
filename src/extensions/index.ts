import { ExtendsActions } from './actions';
import { ExtendsActivityBar } from './activityBar';
import { ExtendsAuxiliaryBar } from './auxiliaryBar';
import { ExtendsContextMenu } from './contextMenu';
import { ExtendsEditor } from './editor';
import { ExtendsEditorTree } from './editorTree';
import { ExtendsExplorer } from './explorer';
import { ExtendsFolderTree } from './folderTree';
import { LocalesExtends } from './locales';
import { ExtendsMenuBar } from './menuBar';
import { ExtendsNotification } from './notification';
import { ExtendsPanel } from './panel';
import { ExtendsSearch } from './search';
import { ExtendsSettings } from './settings';
import { ExtendsSidebar } from './sidebar';
import { ExtendsStatusBar } from './statusBar';
import { ColorThemeExtension } from './themes';

export default [
    LocalesExtends,
    ColorThemeExtension,
    ExtendsActions,
    ExtendsEditor,
    ExtendsActivityBar,
    ExtendsEditorTree,
    ExtendsPanel,
    ExtendsMenuBar,
    ExtendsNotification,
    ExtendsStatusBar,
    ExtendsSearch,
    ExtendsSettings,
    ExtendsFolderTree,
    ExtendsSidebar,
    ExtendsExplorer,
    ExtendsContextMenu,
    ExtendsAuxiliaryBar,
];
