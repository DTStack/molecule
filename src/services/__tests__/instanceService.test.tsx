import React from 'react';
import { defaultExtensions } from 'mo/extensions';
import InstanceService from '../instanceService';
import { render } from '@testing-library/react';
import { ExtendsLocales } from 'mo/extensions/locales-defaults';

describe('The InstanceService', () => {
    test('Constuctor with default config', () => {
        const instance = new InstanceService({}, {} as any);
        const config = instance.getConfig();
        expect(config.defaultLocale).toBe('en');
        expect(config.extensions).toEqual(defaultExtensions);
    });

    test('Should init with params', () => {
        const instance = new InstanceService(
            {
                defaultLocale: 'test',
                extensions: [
                    {
                        id: 1,
                        name: 'test',
                        activate: () => {},
                        dispose: () => {},
                    },
                ],
            },
            {} as any
        );
        const config = instance.getConfig();
        expect(config.defaultLocale).toBe('test');
        expect(config.extensions).toHaveLength(defaultExtensions.length + 1);
    });

    test('Should support render workbench', () => {
        const mockFn = jest.fn();
        const instance = new InstanceService(
            {},
            {
                // @ts-ignore
                extension: {
                    load: jest.fn(),
                    splitLanguagesExts: jest.fn((ext) => [ext, []]),
                },
                // @ts-ignore
                i18n: {
                    initialize: mockFn,
                },
                // @ts-ignore
                monacoService: {
                    initWorkspace: jest.fn(),
                },
                // @ts-ignore
                layout: {
                    container: document.body,
                },
            }
        );
        const { asFragment } = render(instance.render(<div>123</div>));
        expect(asFragment()).toMatchSnapshot();

        expect(mockFn.mock.calls[0][0]).toHaveLength(3);
        expect(mockFn.mock.calls[0][0]).toEqual(
            ExtendsLocales.contributes!.languages
        );
    });

    test('Should support liftCycle hooks', () => {
        const instance = new InstanceService(
            {},
            {
                // @ts-ignore
                extension: {
                    splitLanguagesExts: jest.fn(() => [[], []] as any),
                    load: jest.fn(),
                },
                // @ts-ignore
                i18n: {
                    initialize: jest.fn(),
                },
                // @ts-ignore
                monacoService: {
                    initWorkspace: jest.fn(),
                },
                // @ts-ignore
                layout: {
                    container: document.body,
                },
            }
        );
        const mockFn = jest.fn();
        instance.onBeforeInit(mockFn);
        instance.onBeforeLoad(mockFn);
        instance.render(<div>123</div>);

        expect(mockFn).toBeCalledTimes(2);
    });
});
