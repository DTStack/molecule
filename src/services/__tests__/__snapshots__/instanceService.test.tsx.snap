// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The InstanceService Should support render workbench 1`] = `
<DocumentFragment>
  <div>
    123
  </div>
</DocumentFragment>
`;
